pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/ConstantArcSizeFST.java

```diff
@@ -0,0 +1,159 @@
+package org.apache.pinot.segment.local.utils.nativefst;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import org.apache.pinot.segment.local.utils.nativefst.builders.FSTBuilder;
+
+
+/**
+ * A FST with constant-size arc representation produced directly by
+ * {@link FSTBuilder}.
+ *
+ * @see FSTBuilder
+ */
+public final class ConstantArcSizeFST extends FST {
+  /** Size of the target address field (constant for the builder). */
+  public final static int TARGET_ADDRESS_SIZE = 4;
+
+  /** Size of the flags field (constant for the builder). */
+  public final static int FLAGS_SIZE = 1;
+
+  /** Size of the label field (constant for the builder). */
+  public final static int LABEL_SIZE = 1;
+
+  /**
+   * Size of a single arc structure.
+   */
+  public final static int ARC_SIZE = FLAGS_SIZE + LABEL_SIZE + TARGET_ADDRESS_SIZE;
+
+  /** Offset of the flags field inside an arc. */
+  public final static int FLAGS_OFFSET = 0;
+
+  /** Offset of the label field inside an arc. */
+  public final static int LABEL_OFFSET = FLAGS_SIZE;
+
+  /** Offset of the address field inside an arc. */
+  public final static int ADDRESS_OFFSET = LABEL_OFFSET + LABEL_SIZE;
+  /**
+   * An arc flag indicating the target node of an arc corresponds to a final
+   * state.
+   */
+  public final static int BIT_ARC_FINAL = 1 << 1;
+  /** An arc flag indicating the arc is last within its state. */
+  public final static int BIT_ARC_LAST = 1 << 0;
+  /** A dummy address of the terminal state. */
+  public final static int TERMINAL_STATE = 0;
+  /**
+   * An epsilon state. The first and only arc of this state points either to the
+   * root or to the terminal state, indicating an empty automaton.
+   */
+  private final int _epsilon;
+
+  /**
+   * FST data, serialized as a byte array.
+   */
+  private final byte[] _data;
+
+  private Map<Integer, Integer> _outputSymbols;
+
+  /**
+   * @param data
+   *          FST data. There must be no trailing bytes after the last state.
+   */
+  public ConstantArcSizeFST(byte[] data, int epsilon, Map<Integer, Integer> outputSymbols) {
+    assert epsilon == 0 : "Epsilon is not zero?";
+
+    this._epsilon = epsilon;
+    this._data = data;
+    this._outputSymbols = outputSymbols;
+  }
+
+  @Override
+  public int getRootNode() {
+    return getEndNode(getFirstArc(_epsilon));
+  }
+
+  @Override
+  public int getFirstArc(int node) {
+    return node;
+  }
+
+  @Override
+  public int getArc(int node, byte label) {
+    for (int arc = getFirstArc(node); arc != 0; arc = getNextArc(arc)) {
+      if (getArcLabel(arc) == label) {
+        return arc;
+      }
+    }
+    return 0;
+  }
+
+  @Override
+  public int getNextArc(int arc) {
+    if (isArcLast(arc)) {
+      return 0;
+    }
+    return arc + ARC_SIZE;
+  }
+
+  @Override
+  public byte getArcLabel(int arc) {
+    return _data[arc + LABEL_OFFSET];
+  }
+
+  @Override
+  public int getOutputSymbol(int arc) {
+    return _outputSymbols.get(arc);
```

Comment:
NPE here when `arc` is not in the map.

Reply:
Actually, I did that intentionally. If the arc is not present, then it is a bug.

Reply:
Can you add a comment please?

Reply:
@richardstartin The base method has a callout where it says that a null will be returned if the output symbol is not present.